### ISSUE TITLE
Tab complete for a SET PAYLOAD causes an error if Postgres isn't running

### ISSUE THREAD

**Issue by reybango** [rapid7/metasploit-framework#12076]
## Steps to reproduce

How'd you do it?

1. Typed "set PAYLOAD windows/met" and hit the TAB key
2. Received error that killed Metasploit and took me back to the terminal
3. Reloaded Metasploit and noticed that Postgres wasn't running

This section should also tell us any relevant information about the
environment; for example, if an exploit that used to work is failing,
tell us the victim operating system and service versions.

## Expected behavior

I expected the payload name to be tab completed

## Current behavior

What happens instead?

```
msf5 exploit(multi/browser/java_jre17_provider_skeleton) > set PAYLOAD windows/metTraceback (most recent call last):
	28: from /usr/bin/msfconsole:49:in `<main>'
	27: from /usr/share/metasploit-framework/lib/metasploit/framework/command/base.rb:82:in `start'
	26: from /usr/share/metasploit-framework/lib/metasploit/framework/command/console.rb:48:in `start'
	25: from /usr/share/metasploit-framework/lib/rex/ui/text/shell.rb:134:in `run'
	24: from /usr/share/metasploit-framework/lib/rex/ui/text/shell.rb:309:in `get_input_line'
	23: from /usr/share/metasploit-framework/lib/rex/ui/text/input/readline.rb:100:in `pgets'
	22: from /usr/share/metasploit-framework/lib/rex/ui/text/input/readline.rb:162:in `readline_with_output'
	21: from /usr/share/metasploit-framework/vendor/bundle/ruby/2.5.0/gems/rb-readline-0.5.5/lib/rbreadline.rb:4875:in `readline'
	20: from /usr/share/metasploit-framework/vendor/bundle/ruby/2.5.0/gems/rb-readline-0.5.5/lib/rbreadline.rb:4853:in `readline_internal'
	19: from /usr/share/metasploit-framework/vendor/bundle/ruby/2.5.0/gems/rb-readline-0.5.5/lib/rbreadline.rb:4779:in `readline_internal_charloop'
	18: from /usr/share/metasploit-framework/vendor/bundle/ruby/2.5.0/gems/rb-readline-0.5.5/lib/rbreadline.rb:4363:in `_rl_dispatch'
	17: from /usr/share/metasploit-framework/vendor/bundle/ruby/2.5.0/gems/rb-readline-0.5.5/lib/rbreadline.rb:4374:in `_rl_dispatch_subseq'
	16: from /usr/share/metasploit-framework/vendor/bundle/ruby/2.5.0/gems/rb-readline-0.5.5/lib/rbreadline.rb:6903:in `rl_complete'
	15: from /usr/share/metasploit-framework/vendor/bundle/ruby/2.5.0/gems/rb-readline-0.5.5/lib/rbreadline.rb:6813:in `rl_complete_internal'
	14: from /usr/share/metasploit-framework/vendor/bundle/ruby/2.5.0/gems/rb-readline-0.5.5/lib/rbreadline.rb:6329:in `gen_completion_matches'
	13: from /usr/share/metasploit-framework/vendor/bundle/ruby/2.5.0/gems/rb-readline-0.5.5/lib/readline.rb:136:in `readline_attempted_completion_function'
	12: from /usr/share/metasploit-framework/lib/rex/ui/text/shell.rb:62:in `block in init_tab_complete'
	11: from /usr/share/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:375:in `tab_complete'
	10: from /usr/share/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:390:in `tab_complete_stub'
	 9: from /usr/share/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:390:in `each'
	 8: from /usr/share/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:401:in `block in tab_complete_stub'
	 7: from /usr/share/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:442:in `tab_complete_helper'
	 6: from /usr/share/metasploit-framework/lib/msf/ui/console/command_dispatcher/core.rb:1615:in `cmd_set_tabs'
	 5: from /usr/share/metasploit-framework/lib/msf/ui/console/command_dispatcher/core.rb:2167:in `tab_complete_option'
	 4: from /usr/share/metasploit-framework/lib/msf/ui/console/command_dispatcher/core.rb:2304:in `option_values_payloads'
	 3: from /usr/share/metasploit-framework/lib/msf/core/exploit.rb:775:in `compatible_payloads'
	 2: from /usr/share/metasploit-framework/lib/msf/core/exploit.rb:775:in `each'
	 1: from /usr/share/metasploit-framework/lib/msf/core/exploit.rb:776:in `block in compatible_payloads'
/usr/share/metasploit-framework/lib/msf/core/exploit.rb:745:in `is_payload_compatible?': undefined method `cached_size' for nil:NilClass (NoMethodError)
```

### Metasploit version

Framework: 5.0.35-dev
Console  : 5.0.35-dev

### I installed Metasploit with:
- [X ] Kali package via apt
- [ ] Omnibus installer (nightly)
- [ ] Commercial/Community installer (from http://www.rapid7.com/products/metasploit/download.jsp)
- [ ] Source install (please specify ruby version)

### OS

What OS are you running Metasploit on?

Linux kali 4.19.0-kali5-amd64 #1 SMP Debian 4.19.37-5kali1 (2019-06-20) x86_64 GNU/Linux

**Comment by wvu-r7** [comment]
Strange, I can't reproduce this yet. Likely related to #11768.

What does `show payloads` say?

**Comment by wvu-r7** [comment]
I'm not sure the circumstances where the module cache would return an invalid module, but this is really a bug fix for `is_payload_compatible?`, not necessarily `compatible_payloads`, which merely uses it.

```
msf5 exploit(multi/browser/java_jre17_provider_skeleton) > pry
[*] Starting Pry shell...
[*] You are in exploit/multi/browser/java_jre17_provider_skeleton

[1] pry(#<Msf::Modules::Exploit__Multi__Browser__Java_jre17_provider_skeleton::MetasploitModule>)> compatible_payloads
=> [["generic/custom", #<Class:0x00007fce38229530>],
 ["generic/debug_trap", #<Class:0x00007fce38232a18>],
 ["generic/shell_bind_tcp", #<Class:0x00007fce382315c8>],
 ["generic/shell_reverse_tcp", #<Class:0x00007fce38230e48>],
 ["generic/tight_loop", #<Class:0x00007fce382338c8>],
 ["windows/dllinject/bind_hidden_ipknock_tcp", #<Class:0x00007fce3829b860>],
 ["windows/dllinject/bind_hidden_tcp", #<Class:0x00007fce35b9cf98>],
 ["windows/dllinject/bind_ipv6_tcp", #<Class:0x00007fce38159768>],
 ["windows/dllinject/bind_ipv6_tcp_uuid", #<Class:0x00007fce359e5510>],
 ["windows/dllinject/bind_named_pipe", #<Class:0x00007fce35ae78a0>],
 ["windows/dllinject/bind_nonx_tcp", #<Class:0x00007fce38933830>],
 ["windows/dllinject/bind_tcp", #<Class:0x00007fce380d1cc8>],
 ["windows/dllinject/bind_tcp_rc4", #<Class:0x00007fce347a2290>],
 ["windows/dllinject/bind_tcp_uuid", #<Class:0x00007fce381514c8>],
 ["windows/dllinject/reverse_hop_http", #<Class:0x00007fce345a2620>],
 ["windows/dllinject/reverse_http", #<Class:0x00007fce344a2c48>],
 ["windows/dllinject/reverse_http_proxy_pstore", #<Class:0x00007fce388631f8>],
 ["windows/dllinject/reverse_ipv6_tcp", #<Class:0x00007fce379b8a18>],
 ["windows/dllinject/reverse_nonx_tcp", #<Class:0x00007fce33f42a50>],
 ["windows/dllinject/reverse_ord_tcp", #<Class:0x00007fce3884a590>],
 ["windows/dllinject/reverse_tcp", #<Class:0x00007fce382318e8>],
 ["windows/dllinject/reverse_tcp_allports", #<Class:0x00007fce31eefa00>],
 ["windows/dllinject/reverse_tcp_dns", #<Class:0x00007fce38040700>],
 ["windows/dllinject/reverse_tcp_rc4", #<Class:0x00007fce381a8138>],
 ["windows/dllinject/reverse_tcp_rc4_dns", #<Class:0x00007fce34718450>],
 ["windows/dllinject/reverse_tcp_uuid", #<Class:0x00007fce31f8f578>],
 ["windows/dllinject/reverse_winhttp", #<Class:0x00007fce33d46cb0>],
 ["windows/dns_txt_query_exec", #<Class:0x00007fce38208ec0>],
 ["windows/download_exec", #<Class:0x00007fce381fa8c0>],
 ["windows/exec", #<Class:0x00007fce38200cc0>],
 ["windows/loadlibrary", #<Class:0x00007fce3820b968>],
 ["windows/messagebox", #<Class:0x00007fce381f91a0>],
 ["windows/meterpreter/bind_hidden_ipknock_tcp", #<Class:0x00007fce38282158>],
 ["windows/meterpreter/bind_hidden_tcp", #<Class:0x00007fce35b7fa10>],
 ["windows/meterpreter/bind_ipv6_tcp", #<Class:0x00007fce35b4dfb0>],
 ["windows/meterpreter/bind_ipv6_tcp_uuid", #<Class:0x00007fce38012f58>],
 ["windows/meterpreter/bind_named_pipe", #<Class:0x00007fce35abe1d0>],
 ["windows/meterpreter/bind_nonx_tcp", #<Class:0x00007fce38912b08>],
 ["windows/meterpreter/bind_tcp", #<Class:0x00007fce380a98b8>],
 ["windows/meterpreter/bind_tcp_rc4", #<Class:0x00007fce347814c8>],
 ["windows/meterpreter/bind_tcp_uuid", #<Class:0x00007fce3813a638>],
 ["windows/meterpreter/reverse_hop_http", #<Class:0x00007fce33ea8fb8>],
 ["windows/meterpreter/reverse_http", #<Class:0x00007fce3590df98>],
 ["windows/meterpreter/reverse_http_proxy_pstore", #<Class:0x00007fce37a51628>],
 ["windows/meterpreter/reverse_https", #<Class:0x00007fce35cad810>],
 ["windows/meterpreter/reverse_https_proxy", #<Class:0x00007fce34274700>],
 ["windows/meterpreter/reverse_ipv6_tcp", #<Class:0x00007fce359d7eb0>],
 ["windows/meterpreter/reverse_named_pipe", #<Class:0x00007fce340e4e30>],
 ["windows/meterpreter/reverse_nonx_tcp", #<Class:0x00007fce33f22de0>],
 ["windows/meterpreter/reverse_ord_tcp", #<Class:0x00007fce3882a240>],
 ["windows/meterpreter/reverse_tcp", #<Class:0x00007fce38203038>],
 ["windows/meterpreter/reverse_tcp_allports", #<Class:0x00007fce31ecf868>],
 ["windows/meterpreter/reverse_tcp_dns", #<Class:0x00007fce35a3fd08>],
[snip]
[2] pry(#<Msf::Modules::Exploit__Multi__Browser__Java_jre17_provider_skeleton::MetasploitModule>)> is_payload_compatible?(nil)
NoMethodError: undefined method `cached_size' for nil:NilClass
from /rapid7/metasploit-framework/lib/msf/core/exploit.rb:745:in `is_payload_compatible?'
[3] pry(#<Msf::Modules::Exploit__Multi__Browser__Java_jre17_provider_skeleton::MetasploitModule>)> is_payload_compatible?('')
NoMethodError: undefined method `cached_size' for nil:NilClass
from /rapid7/metasploit-framework/lib/msf/core/exploit.rb:745:in `is_payload_compatible?'
[4] pry(#<Msf::Modules::Exploit__Multi__Browser__Java_jre17_provider_skeleton::MetasploitModule>)> is_payload_compatible?('windows/met')
NoMethodError: undefined method `cached_size' for nil:NilClass
from /rapid7/metasploit-framework/lib/msf/core/exploit.rb:745:in `is_payload_compatible?'
[5] pry(#<Msf::Modules::Exploit__Multi__Browser__Java_jre17_provider_skeleton::MetasploitModule>)> is_payload_compatible?('windows/meterpreter/reverse_tcp')
=> true
[6] pry(#<Msf::Modules::Exploit__Multi__Browser__Java_jre17_provider_skeleton::MetasploitModule>)>
```

**Comment by wvu-r7** [comment]
Reopening this until we can confirm root cause.

**Comment by reybango** [comment]
@wvu-r7 tried to dupe it again and can't replicate the issue.

**Comment by wvu-r7** [comment]
Gah, that's even worse, haha. We're about to test in Kali. For now, #12085 is merged in the tree and will be shipped out to Kali in a week. It should prevent the crash, but it might mean no compatible payloads exist for another more awful reason.

**Comment by jmartin-r7** [comment]
Local test on Kali version from apt does not reproduce, any tab completion attempt for incompatible payload gets no output.

```
msf5 exploit(multi/browser/java_jre17_provider_skeleton) > set PAYLOAD
set PAYLOAD generic/custom                   set PAYLOAD java/jsp_shell_reverse_tcp       set PAYLOAD java/meterpreter/reverse_tcp     set PAYLOAD multi/meterpreter/reverse_http
set PAYLOAD generic/shell_bind_tcp           set PAYLOAD java/meterpreter/bind_tcp        set PAYLOAD java/shell/bind_tcp              set PAYLOAD multi/meterpreter/reverse_https
set PAYLOAD generic/shell_reverse_tcp        set PAYLOAD java/meterpreter/reverse_http    set PAYLOAD java/shell/reverse_tcp
set PAYLOAD java/jsp_shell_bind_tcp          set PAYLOAD java/meterpreter/reverse_https   set PAYLOAD java/shell_reverse_tcp
msf5 exploit(multi/browser/java_jre17_provider_skeleton) > set PAYLOAD windo
```

**Comment by wvu-r7** [comment]
Symptom fixed in #12085. Will wait to see if the root cause is reproducible.
